### PR TITLE
Cleanup previews in the background

### LINKED ISSUE #5263: [Previews][Question/Request] Obsolete previews seem to be not cleaned regularly
In relation to the question of @JSoko in previewgenerator: https://github.com/rullzer/previewgenerator/issues/64

I recreated the issue here, because it's actually a question about the core preview system:
- Are the file previews regularly checked for obsolete/orphaned ones and cleaned like/together with the filecache database?
- If not, it would be helpful to have a command + background job the regularly do that.
- Normally previews will be removed automatically, if their related files are removed in common way, done by the [watcher.php](https://github.com/nextcloud/server/blob/master/lib/private/Preview/Watcher.php). But for process errors, manual file deletion (outside of nextcloud) and other reasons the deletion could be not tracked by nextcloud/watcher and the preview could remain and forgotten, consuming capacity.
- For the same reason there is the `occ files:cleanup` command and a related background job to clean the filecache/index. It seems just reasonable to have the same for previews or simply include previews into the existing jobs.
- Excuse me, if this already happens, but I didn't found it in the code so far and [files:cleanup](https://github.com/nextcloud/server/blob/master/apps/files/lib/BackgroundJob/DeleteOrphanedItems.php) seems to just include tags and comments:
```
	public function run($argument) {
		$this->cleanSystemTags();
		$this->cleanUserTags();
		$this->cleanComments();
		$this->cleanCommentMarkers();
	}
```

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #5263 

* Make the AppData a bit more strict
* Add a backgroundjob that cleans up previews
  - Runs max once an hour
  - max 10 folders to cleanup if no cli cron
  - Else just loop over all
* Remove the live preview cleanup
  - This easily blows up number of files to remove if you delete a large folder
  - As a bonus now the previews stay there as long as the file is still in the trash

### CODE REVIEW

**Comment by rullzer** [comment]
@MorrisJobke fixed

**Comment by icewind1991** [comment]
The query fails on the comparison between `fileid` and `name` on postgresql without an explicit type cast.

And please add unit tests.

(I also think that the query can be a lot more efficient using joins instead of subqueries, will do some experimenting)

**Comment by rullzer** [comment]
@icewind1991 aaah good catch... I was atually thinking that yesterday evening that this might happen... I'll take care

**Comment by rullzer** [comment]
@icewind1991 good call.
Updated and now with tests.

File: tests/lib/Preview/BackgroundCleanupJobTest.php [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	/** @var PreviewManager */
+	private $previewManager;
+
+	/** @var IRootFolder */
+	private $rootFolder;
+
+	public function setUp() {
+		parent::setUp();
+
+		$storage = new \OC\Files\Storage\Temporary([]);
```
Comment (icewind1991) [inline tests/lib/Preview/BackgroundCleanupJobTest.php]: `MountProviderTrait` and `UserTrait` that provide helpers for setting up users and mounts for tests

Comment (rullzer) [reply tests/lib/Preview/BackgroundCleanupJobTest.php]: Done

**Comment by rullzer** [comment]
Aaaah I'm to greedy here. Of course the watcher needs to stay to invalidate previews on write... I'll fix that

**Comment by rullzer** [comment]
Ok lets do this :)

**Review by icewind1991 — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Pull request merged**